pynisher/pynisher.py

```
@@ -0,0 +1,450 @@
+
+        # Let loose
+        subprocess.start()
+
+        # If self.wall time is None, block until the subprocess finishes or terminates
+        # Otherwise, will return after wall_time and the process will still be running
+        subprocess.join(self.wall_time)
+
+        # exitcode here can only take on 3 values
+        #
+        # * None        | the subprocess is still running (walltime elapsed)
+        # * 0           | Ended gracefully, pipe is closed or response in the pipe
+        # * >0          | Process was terminated non-gracefully, nothing in the pipe
+        #               | and it may not be closed
+        exitcode = subprocess.exitcode
+
+        # Wall time expired
+        if exitcode is None:
+            subprocess.terminate()
+
+            result = EMPTY
+            err = WallTimeoutException(
+                f"Did not finish in time ({self.wall_time}s)"
+                f"\n{callstring(self.func, *args, **kwargs)}"
+            )
+
+        # Completed gracefully
+        elif exitcode == 0:
+
+            # Pipe has data in it or has been closed properly
+            if receive_pipe.poll():
+
+                # If there is data, it will be read
+                try:
+                    response = receive_pipe.recv()
+                    if response is None:
+                        result = EMPTY
+                        err = MemoryLimitException(
+                            "Ended gracefully but only `None` could be sent back"
+                            " from the process. This is a MemoryException as"
+                            " previous errors could not be sent previously."
+                            f"\n{callstring(self.func, *args, **kwargs)}"
+                        )
+                    else:
+                        result, err, tb = response
+                        # If we have an err, that excludes possible results
+                        if err is not None:
+                            result = EMPTY
```

Should we raise an error if there is both an error and a result?

---

That's really not possible as seens from `limiter.py`, I would throw in an assert because it means an implementation issue

---

I can't throw in an assert as `(None, None, None)` is a valid repsonse. This would mean the users function return `None` and there is no error.

We have no way to check if there was a result and an error as `None` is a valid value for a function to return.

The previous comments about `limiter.py` still hold though so I think it's fine.

---

Thanks for the explanation.